File: Dynamic Programming/Hard/LET_no446_Arithmetic Slices II - Subsequence.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        res = 0
        dp = [collections.defaultdict(int) for _ in range(len(nums))]
        
        for i in range(len(nums)):
            for j in range(i):
                diff = nums[i] - nums[j]
                dp[i][diff] += dp[j][diff] + 1
                res += dp[j][diff]
        
        return res

File: Dynamic Programming/Hard/LET_no446_Arithmetic Slices II - Subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        int res = 0;
        unordered_map<long, int> dp[nums.size()];
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                long diff = (long)nums[i] - nums[j];
                auto it = dp[j].find(diff);
                int count = it == dp[j].end() ? 0 : it->second;
                res += count;
                dp[i][diff] += count + 1;
            }
        }
        return res;
    }
};